### PR TITLE
Ability to write out package manifests as source code (for mechanical package generation)

### PR DESCRIPTION
**Pull request by abertelrud**
Ability to write out package manifests as source code, providing more flexibility for things like mechanical package manifest generation.

Note that this is _not_ suitable for editing the package manifest, since that would need to preserve whitespace and comments.  Rather, this is for things like automatically generating packages from test sets, or for generating the initial configuration of packages (e.g. a more flexible `package init`).

rdar://70271810

### CODE REVIEW

File: Sources/PackageModel/ManifestSourceGeneration.swift [hunk 1]
```diff
@@ -0,0 +1,326 @@
...
+/// algorithmically generated manifest content, so it is not suitable for the
+/// mechanical editing of package manifests.  Rather, it is intended for such
+/// tasks as manifest creation as part of package instantiation, etc.
+extension Manifest {
+    
+    /// Generates and returns a string containing the contents of the manifest
+    /// in canonical declarative form.
+    public var generatedManifestFileContents: String {
+        return """
+            // swift-tools-version: \(toolsVersion.description)
```
Comment (abertelrud) [inline Sources/PackageModel/ManifestSourceGeneration.swift:26]: Using `description` isn't good enough here; I forgot to add a FIXME in the code.

Comment (WowbaggersLiquidLunch) [reply Sources/PackageModel/ManifestSourceGeneration.swift:26]: I think just toolsVersion should be enough. I'm not completely sure, but I think since ToolsVersion conforms to CustomStringConvertible, its instances can be interpolated directly.

Additionally, I'm wondering if this could use toolsVersionWriter? toolsVersionWriter currently prepends the "swift-tools-version" line to the manifest (or replace the line when the tools version is invalid), so it seems like it might be a good fit here?

File: Sources/PackageModel/ManifestSourceGeneration.swift [hunk 2]
```diff
@@ -0,0 +1,326 @@
...
+    }
+}
+
+
+
+/// Helper type to emit source code.  Represents one node of source code, as a
+/// arbitrary string followed by an optional child list, optionally enclosed in
+/// a pair of delimiters.  The code generation works by creating source code
+/// fragments and then rendering them as source code with proper formatting.
+fileprivate class SourceCodeFragment {
```
Comment (tomerd) [inline Sources/PackageModel/ManifestSourceGeneration.swift:266]: is this actually carrying state and needs to be a class?

Comment (abertelrud) [reply Sources/PackageModel/ManifestSourceGeneration.swift:266]: I'd prefer a struct, but since it can have children of the same type, I got errors.  I think there's a way to have a by-reference struct, though, so I was going to make that change before making this no longer be a draft.

File: Sources/PackageModel/ManifestSourceGeneration.swift (hunk 1 above)
Comment (abertelrud) [reply Sources/PackageModel/ManifestSourceGeneration.swift:26]: I wasn't aware of that, but it sounds like a good fit.  I'll look into it.  Thanks!

File: Sources/PackageModel/ManifestSourceGeneration.swift (hunk 2 above)
Comment (abertelrud) [reply Sources/PackageModel/ManifestSourceGeneration.swift:266]: I think `indirect` is what I might be looking for.  Will try that.

File: Sources/PackageModel/ManifestSourceGeneration.swift (hunk 1 above)
Comment (abertelrud) [reply Sources/PackageModel/ManifestSourceGeneration.swift:26]: `toolsVersionWriter` looks promising but seems to rely on a file system.  If this were split out into a ToolsVersionFormatter, that would be perfect, and `toolsVersionWriter` were written in terms of that, that would be great.

Comment (abertelrud) [reply Sources/PackageModel/ManifestSourceGeneration.swift:26]: Also, the `Format.asJSON(version.major)` and `Format.asJSON(version.minor)` look a little odd.  These version numbers are not quoted or processed in other ways (and there is no connection to JSON), so that code looks a little suspicious.

Comment (WowbaggersLiquidLunch) [reply Sources/PackageModel/ManifestSourceGeneration.swift:26]: > If this were split out into a ToolsVersionFormatter, that would be perfect, and `toolsVersionWriter` were written in terms of that, that would be great.

This makes sense, and I kind of want to take it on myself once #2937 (which modifies the file) settles. 

> Also, the `Format.asJSON(version.major)` and `Format.asJSON(version.minor)` look a little odd. These version numbers are not quoted or processed in other ways (and there is no connection to JSON), so that code looks a little suspicious.

Yes `asJSON` is quite odd. [It actually doesn't do anything.](https://github.com/apple/swift-tools-support-core/blob/8123987b1681af3db7b695d05627d58c979809f3/Sources/TSCBasic/WritableByteStream.swift#L445) So it's probably safe to just replace it with string interpolation.

**Comment by abertelrud** [comment]
There are a couple of edge-case FIXMEs to address, then this is ready.

**Comment by abertelrud** [comment]
@swift-ci please smoke test

**Comment by abertelrud** [comment]
@swift-ci please smoke test

**Comment by abertelrud** [comment]
@swift-ci please smoke test

**Comment by abertelrud** [comment]
@swift-ci please smoke test

**Review by neonichu — APPROVED** [review]
